src/core/stdcpp/typeinfo.d

```diff
@@ -0,0 +1,146 @@
+// Written in the D programming language.
+
+/**
+ * Interface to C++ <typeinfo>
+ *
+ * Copyright: Copyright (c) 2016 D Language Foundation
+ * License:   $(WEB boost.org/LICENSE_1_0.txt, Boost License 1.0).
+ * Authors:   $(WEB digitalmars.com, Walter Bright)
+ * Source:    $(DRUNTIMESRC core/stdcpp/_typeinfo.d)
+ */
+
+module core.stdcpp.typeinfo;
+
+version (CRuntime_DigitalMars)
+{
+    import core.stdcpp.exception;
+
+    extern (C++, std)
+    {
+        class type_info
+        {
+            void* pdata;
+
+          public:
+            //virtual ~this();
+            void dtor() { }     // reserve slot in vtbl[]
+
+            //bool operator==(const type_info rhs) const;
+            //bool operator!=(const type_info rhs) const;
+            final bool before(const type_info rhs) const;
+            final const(char)* name() const;
+          protected:
+            //type_info();
+          private:
+            //this(const type_info rhs);
+            //type_info operator=(const type_info rhs);
+        }
+
+        class bad_cast : core.stdcpp.exception.std.exception
+        {
+            this() nothrow { }
+            this(const bad_cast) nothrow { }
+            //bad_cast operator=(const bad_cast) nothrow { return this; }
+            //virtual ~this() nothrow;
+            override const(char)* what() const nothrow;
+        }
+
+        class bad_typeid : core.stdcpp.exception.std.exception
+        {
+            this() nothrow { }
+            this(const bad_typeid) nothrow { }
+            //bad_typeid operator=(const bad_typeid) nothrow { return this; }
+            //virtual ~this() nothrow;
+            override const (char)* what() const nothrow;
+        }
+    }
+}
+else version (CRuntime_Microsoft)
+{
+    import core.stdcpp.exception;
+
+    struct __type_info_node
+    {
+        void* _MemPtr;
+        __type_info_node* _Next;
+    }
+
+    extern __gshared __type_info_node __type_info_root_node;
+
+    extern (C++, std)
+    {
+        class type_info
+        {
+
+          public:
+            //virtual ~this();
+            void dtor() { }     // reserve slot in vtbl[]
+            //bool operator==(const type_info rhs) const;
+            //bool operator!=(const type_info rhs) const;
+            final bool before(const type_info rhs) const;
+            final const(char)* name(__type_info_node* p = &__type_info_root_node) const;
+
+          private:
+            void* pdata;
+            char _name[1];
+            //this(const type_info rhs);
+            //type_info operator=(const type_info rhs);
+        }
+
+        class bad_cast : core.stdcpp.exception.std.exception
+        {
+            this(const(char)* msg = "bad cast") { }
```

Comment:
>if it did then it would corrupt memory and/or crash.

Looks like the "complete object constructor/destructor" (C1/D1 mangling) has the similar signature with D dtor/ctor: `abi::__cxa_cdtor_return_type (*)(void*)` where __cxa_cdtor_return_type is `void` or `void*`. Thus we should may possibility to call C++ ctor as D ctor:
```
void* __d_create_cpp_object(size_t size, void function(void*) ctor)
{
    auto obj = GC.calloc(size);
    //blitting init[] is possible but isn't needed.
    ctor(obj);
    return obj;
}
```